File: tiendapome.backend/tiendapome.Entidades/DocumentoVenta.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace tiendapome.Entidades
{
    [JsonObject]
    public class DocumentoVenta : EntidadBase
    {        
        [JsonProperty("IdEmpresa")]
        public virtual int IdEmpresa { get; set; }
        [JsonProperty("Usuario")]
        public virtual Cliente Usuario { get; set; }
        [JsonProperty("Cliente")]
        public virtual Cliente Cliente { get; set; }
        [JsonProperty("IdPedido")]
        public virtual int IdPedido { get; set; }
        [JsonProperty("NumeroPedido")]
        public virtual int NumeroPedido { get; set; }
        [JsonProperty("TipoComprobante")]
        public virtual VentaTipoComprobante TipoComprobante { get; set; }        
        [JsonProperty("Letra")]
        public virtual decimal Letra { get; set; }
        [JsonProperty("Sucursal")]
        public virtual int Sucursal { get; set; }
        [JsonProperty("Numero")]
        public virtual int Numero { get; set; }
        [JsonProperty("Fecha")]
        public virtual DateTime Fecha { get; set; }
        [JsonProperty("Vencimiento")]
        public virtual DateTime Vencimiento { get; set; }
        [JsonProperty("Gravado")]
        public virtual decimal Gravado { get; set; }
        [JsonProperty("Descuento")]
        public virtual decimal Descuento { get; set; }
        [JsonProperty("PorcentajeIVA")]
        public virtual decimal PorcentajeIVA { get; set; }
        [JsonProperty("IVA")]
        public virtual decimal IVA { get; set; }
        [JsonProperty("Total")]
        public virtual decimal Total { get; set; }
        [JsonProperty("Pendiente")]
        public virtual decimal Pendiente { get; set; }
        [JsonProperty("Comision")]
        public virtual decimal Comision { get; set; }
        
        [JsonProperty("Efectivo")]
        public virtual decimal Efectivo { get; set; }
        [JsonProperty("EfectivoCotizaDolar")]
        public virtual decimal EfectivoCotizaDolar { get; set; }
        
        [JsonProperty("Dolares")]
        public virtual decimal Dolares { get; set; }
        [JsonProperty("DolaresCotizaDolar")]
        public virtual decimal DolaresCotizaDolar { get; set; }
        
        [JsonProperty("Euros")]
        public virtual decimal Euros { get; set; }
        [JsonProperty("EurosCotizaDolar")]
        public virtual decimal EurosCotizaDolar { get; set; }
        
        [JsonProperty("Cheques")]
        public virtual decimal Cheques { get; set; }
        [JsonProperty("ChequesCotizaDolar")]
        public virtual decimal ChequesCotizaDolar { get; set; }
        
        [JsonProperty("Tarjeta")]
        public virtual decimal Tarjeta { get; set; }
        [JsonProperty("TarjetaCotizaDolar")]
        public virtual decimal TarjetaCotizaDolar { get; set; }
        
        [JsonProperty("MercadoPago")]
        public virtual decimal MercadoPago { get; set; }
        [JsonProperty("MercadoPagoCotizaDolar")]
        public virtual decimal MercadoPagoCotizaDolar { get; set; }
        
        [JsonProperty("DepositoTransferencia")]
        public virtual decimal DepositoTransferencia { get; set; }
        [JsonProperty("DepositoTransferCotizaDolar")]
        public virtual decimal DepositoTransferCotizaDolar { get; set; }
        
        [JsonProperty("RetencionIVA")]
        public virtual decimal RetencionIVA { get; set; }
        [JsonProperty("RetencionGanancia")]
        public virtual decimal RetencionGanancia { get; set; }
        [JsonProperty("RetencionIngBrutos")]
        public virtual decimal RetencionIngBrutos { get; set; }
        [JsonProperty("Anulado")]
        public virtual bool Anulado { get; set; }

        [JsonProperty("Items")]
        public virtual List<DocumentoVentaItem> Items { get; set; }

        [JsonProperty("Observaciones")]
        public virtual DocumentoVentaObservaciones Observaciones{ get; set; }


        [JsonProperty("IdTipoComprobanteGenerar")]
        public virtual int IdTipoComprobanteGenerar { get; set; }

        [JsonProperty("DiasVencido")]
        public virtual int DiasVencido 
        {
            get 
            {
                TimeSpan tSpan = DateTime.Now - this.Vencimiento;
                return this.Pendiente == 0 ? 0 : tSpan.Days;
            }
            set { }
        }

        public DocumentoVenta() { }

        public void Validar()
        {
            StringBuilder sb = new StringBuilder();

            if (Cliente == null)
                sb.AppendLine("Debe indicar cliente.");

            if (Usuario == null)
                sb.AppendLine("Debe indicar usuario.");

            if (TipoComprobante == null)
                sb.AppendLine("Debe seleccionar un tipo de comprobante.");

        }
    }
}

File: tiendapome.frontend/src/app/models/documentoVenta.ts
import { Cliente } from './cliente';

export class DocumentoVenta {
    Id: number;
    IdEmpresa: number;
    Usuario: Cliente;
    Cliente: Cliente;
    IdPedido: number;
    NumeroPedido: number;
    TipoComprobante: VentaTipoComprobante;
    Letra: string;
    Sucursal: number;
    Numero: number;
    Fecha: Date;
    Vencimiento: Date;
    Gravado: number;
    Descuento: number;
    PorcentajeIVA: number;
    IVA: number;
    Total: number;
    Pendiente: number;
    Comision: number;
    Efectivo: number;
    EfectivoCotizaDolar: number;
    Dolares: number;
    DolaresCotizaDolar: number;
    Euros: number;
    EurosCotizaDolar: number;
    Cheques: number;
    ChequesCotizaDolar: number;
    Tarjeta: number;
    TarjetaCotizaDolar: number;
    MercadoPago: number;
    MercadoPagoCotizaDolar: number;
    DepositoTransferencia: number;
    DepositoTransferCotizaDolar: number;
    RetencionIVA: number;
    RetencionGanancia: number;
    RetencionIngBrutos: number;
    Anulado: boolean;
    Items: Array<DocumentoVentaItem>;
    Observaciones: DocumentoVentaObservaciones;
    
    DiasVencido:number;

    Debe: number;
    Haber: number;
    Saldo: number;

    constructor() {
        this.Id = -1;
        this.Usuario = new Cliente();
        this.Cliente = new Cliente();
        this.TipoComprobante = new VentaTipoComprobante();
        this.Items = new Array<DocumentoVentaItem>();
        this.Observaciones = new DocumentoVentaObservaciones();
        this.Numero = 0;
        this.Gravado = 0;
        this.Descuento = 0;
        this.Total = 0;
        this.Fecha = new Date();
        this.Vencimiento = new Date();

    }
}


export class DocumentoVentaItem {
    Id: number;
    IdVenta: number;
    NroItem: number;
    IdProductoStock: number;
    IdPedidoItemProducto: number;
    Descripcion: string;
    Cantidad: number;
    PrecioUnitario: number;
    Precio: number;

    constructor() {
        this.Id = -1;
        this.IdVenta = -1;
        this.NroItem = -1;
    }
}


export class DocumentoVentaObservaciones {
    Id: number;
    IdVenta: number;
    Observaciones: string;
    Adjunto: string;
    AdjuntoLink: string;

    constructor() {
        this.Id = -1;
        this.IdVenta = -1;
    }
}


export class DocumentoVentaList {
    TotalFilas: number;
    DocumentosVenta: DocumentoVenta[];
    constructor() { }
}


export class VentaTipoComprobante {
    Id: number;
    Descripcion: string;
    Vigente: boolean;
    Abreviado: string;
    EsDebe: boolean;

    constructor() {
        this.Id = -1;
    }
}
